controllers/storagecluster/volumegroupsnapshotterclasses.go

```diff
@@ -0,0 +1,169 @@
+					return err
+				}
+				// no error, continue with the next iteration
+				continue
+			}
+
+			r.Log.Error(err, "Failed to 'Get' GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef(vsc.Namespace, vsc.Name))
+			return err
+		}
+		if vscc.reconcileStrategy == ReconcileStrategyInit {
+			return nil
+		}
+		if existing.DeletionTimestamp != nil {
+			return fmt.Errorf("failed to restore GroupSnapshotClass %q because it is marked for deletion", existing.Name)
+		}
+		// if there is a mismatch in the parameters of existing vs created resources,
+		if !reflect.DeepEqual(vsc.Parameters, existing.Parameters) {
+			// we have to update the existing SnapshotClass
+			r.Log.Info("GroupSnapshotClass needs to be updated", "GroupSnapshotClass", klog.KRef(existing.Namespace, existing.Name))
+			existing.ObjectMeta.OwnerReferences = vsc.ObjectMeta.OwnerReferences
+			vsc.ObjectMeta = existing.ObjectMeta
+			if err := r.Client.Update(context.TODO(), vsc); err != nil {
+				r.Log.Error(err, "GroupSnapshotClass updation failed.", "GroupSnapshotClass", klog.KRef(existing.Namespace, existing.Name))
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (obj *ocsGroupSnapshotClass) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	vgsc := newGroupSnapshotClassConfigurations(instance)
+
+	err := r.createGroupSnapshotClasses(vgsc)
+	if err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (obj *ocsGroupSnapshotClass) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	vgscs := newGroupSnapshotClassConfigurations(instance)
+	for _, vgsc := range vgscs {
+		sc := vgsc.groupSnapshotClass
+		existing := groupsnapapi.VolumeGroupSnapshotClass{}
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: sc.Name, Namespace: sc.Namespace}, &existing)
+
+		switch {
+		case err == nil:
+			if existing.DeletionTimestamp != nil {
+				r.Log.Info("Uninstall: GroupSnapshotClass is already marked for deletion.", "GroupSnapshotClass", klog.KRef(existing.Namespace, existing.Name))
+				break
+			}
+
+			r.Log.Info("Uninstall: Deleting GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef(existing.Namespace, existing.Name))
+			existing.ObjectMeta.OwnerReferences = sc.ObjectMeta.OwnerReferences
+			sc.ObjectMeta = existing.ObjectMeta
+
+			err = r.Client.Delete(context.TODO(), sc)
+			if err != nil {
+				r.Log.Error(err, "Uninstall: Ignoring error deleting the GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef(existing.Namespace, existing.Name))
+			}
+		case errors.IsNotFound(err):
+			r.Log.Info("Uninstall: GroupSnapshotClass not found, nothing to do.", "GroupSnapshotClass", klog.KRef(sc.Namespace, sc.Name))
+		default:
+			r.Log.Error(err, "Uninstall: Error while getting GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef(sc.Namespace, sc.Name))
```

Comment:
dont we need to retry if there is any deletion error? i see we are returning nil at the end of this function.

Reply:
Yes adding that in the switch case